### PR TITLE
HitFinder module converted to Shared module, the implementations of tools and algorithms used in this module are thread safe. 

### PR DESCRIPTION
**Pull request by sabasehrish**

### CODE REVIEW

**Comment by yangtj207** [comment]
trigger build

**Comment by yangtj207** [comment]
Lots of failures in the DUNE CI tests:
[https://dbweb5.fnal.gov:8443/LarCI/app/ns:dune/build_detail/phase_details?build_id=dune_ci/9003&platform=Linux%203.10.0-1062.12.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e19:prof](url)

38: %MSG-s ArtException:  PostEndJob 11-May-2020 12:00:29 CDT ModuleEndJob
239: ---- OtherArt BEGIN
240:   An exception was thrown while processing module GausHitFinder/gaushit during beginJob
241:   ---- LogicError BEGIN
242:     An error occurred while processing scheduling options for a module.
243:     Either 'async<art::InEvent>()' or 'serialize<art::InEvent>(...)'
244:     must be called in a shared module's constructor.
245:   ---- LogicError END
246: ---- OtherArt END
247: %MSG

**Comment by absolution1** [comment]
SBND CI unit tests report a similar error:

https://dbweb5.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/test_details?build_id=sbnd_ci/4756&platform=Linux%203.10.0-1062.12.1.el7.x86_64&buildtype=slf7%20e19:prof&phase=unit_test&test=prodsingle_reconstruction_test.sh

18: %MSG-s ArtException:  PostEndJob 11-May-2020 11:54:22 CDT ModuleEndJob
19: ---- OtherArt BEGIN
20:   An exception was thrown while processing module GausHitFinder/gaushit during beginJob
21:   ---- LogicError BEGIN
22:     An error occurred while processing scheduling options for a module.
23:     Either 'async<art::InEvent>()' or 'serialize<art::InEvent>(...)'
24:     must be called in a shared module's constructor.
25:   ---- LogicError END
26: ---- OtherArt END
27: %MSG

**Comment by absolution1** [comment]
trigger build

**Comment by absolution1** [comment]
SBND CI now reports all OK.  I see the DUNE CI reports warnings.  I assume they'll be benign reference updates but I'll hold off proceeding until @yangtj207 or @chalt007 has confirmed this.

File: larreco/HitFinder/GausHitFinder_module.cc [hunk 1]
```diff
@@ -102,30 +109,32 @@ class GausHitFinder : public art::EDProducer {
...
-    fFillHists           = pset.get< bool        >("FillHists",false);
+GausHitFinder::GausHitFinder(fhicl::ParameterSet const& pset, art::ProcessingFrame const&)
+  : SharedProducer{pset},
+    fFilterHits(pset.get< bool>("FilterHits",false)),
+    fFillHists(pset.get<bool>("FillHists",false)),
+    fCalDataModuleLabel(pset.get< std::string >("CalDataModuleLabel")),
+    fAllHitsInstanceName(pset.get< std::string >("AllHitsInstanceName","")),
+    fLongMaxHitsVec(pset.get< std::vector<int>>("LongMaxHits",    std::vector<int>() = {25,25,25})),
+    fLongPulseWidthVec(pset.get< std::vector<int>>("LongPulseWidth", std::vector<int>() = {16,16,16})),
+    fMaxMultiHit(pset.get< int             >("MaxMultiHit")),
```
Comment (absolution1) [inline larreco/HitFinder/GausHitFinder_module.cc:120]: Are the default values needed?
I'd imagine they can only be relevant for one experiment. 
If an experiment misconfigures their fcl file then that experiment would be exposed to these default values.

File: larreco/HitFinder/GausHitFinder_module.cc [hunk 2]
```diff
@@ -102,30 +109,32 @@ class GausHitFinder : public art::EDProducer {
...
+    fAreaMethod(pset.get< int             >("AreaMethod")),
+    fAreaNormsVec(FillOutHitParameterVector(pset.get< std::vector<double> >("AreaNorms"))),
+    fChi2NDF(pset.get< double          >("Chi2NDF")),
+    fPulseHeightCuts(pset.get< std::vector<float>>("PulseHeightCuts", std::vector<float>() = {3.0,  3.0,  3.0})),
+    fPulseWidthCuts(pset.get< std::vector<float>>("PulseWidthCuts",  std::vector<float>() = {2.0,  1.5,  1.0})),
+    fPulseRatioCuts(pset.get< std::vector<float>>("PulseRatioCuts",  std::vector<float>() = {0.35, 0.40, 0.20}))
```
Comment (absolution1) [inline larreco/HitFinder/GausHitFinder_module.cc:126]: Same comment about default values

File: larreco/HitFinder/HitFinderTools/CandHitMorphological_tool.cc [hunk 3]
```diff
@@ -110,32 +110,31 @@ class CandHitMorphological : ICandidateHitFinder
...
+    fDilationFraction   (pset.get< float  >("DilationFraction",    0.75)),
+    fErosionFraction    (pset.get< float  >("ErosionFraction",     0.2)),
+    fMinDeltaTicks      (pset.get< int    >("MinDeltaTicks",       0)),
+    fMinDeltaPeaks      (pset.get< float  >("MinDeltaPeaks",       0.025)),
+    fMinHitHeight       (pset.get< float  >("MinHitHeight",        1.0)),
+    fNumInterveningTicks(pset.get< size_t >("NumInterveningTicks", 6)),
+    fStructuringElement (pset.get< int    >("StructuringElement",  20)),
+    fOutputHistograms   (pset.get< bool   >("OutputHistograms",    false)),
+    fOutputWaveforms    (pset.get< bool   >("OutputWaveforms",     false)),
+    fFitNSigmaFromCenter(pset.get< float  >("FitNSigmaFromCenter", 5.))
```
Comment (absolution1) [inline larreco/HitFinder/HitFinderTools/CandHitMorphological_tool.cc:125]: default values

File: larreco/HitFinder/HitFinderTools/CandHitStandard_tool.cc [hunk 4]
```diff
@@ -41,30 +37,16 @@ class CandHitStandard : ICandidateHitFinder
...
-
-void CandHitStandard::configure(const fhicl::ParameterSet& pset)
-{
-    // Start by recovering the parameters
-    fPlane        = pset.get< int   >("Plane",        0);
-    fRoiThreshold = pset.get< float >("RoiThreshold", 5.);
-
-    return;
+CandHitStandard::CandHitStandard(const fhicl::ParameterSet& pset):
+    fRoiThreshold(pset.get< float >("RoiThreshold", 5.))
```
Comment (absolution1) [inline larreco/HitFinder/HitFinderTools/CandHitStandard_tool.cc:48]: default value

File: larreco/HitFinder/HitFinderTools/PeakFitterGaussian_tool.cc [hunk 5]
```diff
@@ -97,25 +93,14 @@ class PeakFitterGaussian : IPeakFitter
...
-}
-
-PeakFitterGaussian::~PeakFitterGaussian()
+PeakFitterGaussian::PeakFitterGaussian(const fhicl::ParameterSet& pset):
+    fMinWidth(pset.get<double>("MinWidth",         0.5)),
+    fMaxWidthMult (pset.get<double>("MaxWidthMult",     3.)),
+    fPeakRange(pset.get<double>("PeakRangeFact",    2.)),
+    fAmpRange(pset.get<double>("PeakAmpRange",     2.)),
+    fFloatBaseline(pset.get< bool >("FloatBaseline",    false)),
+    fOutputHistograms(pset.get< bool >("OutputHistograms", false))
```
Comment (absolution1) [inline larreco/HitFinder/HitFinderTools/PeakFitterGaussian_tool.cc:102]: default values

File: larreco/HitFinder/HitFinderTools/PeakFitterMrqdt_tool.cc [hunk 6]
```diff
@@ -51,27 +46,15 @@ namespace reco_tool
...
-    fPeakRange = pset.get<double>("PeakRangeFact", 2.);
-    fAmpRange = pset.get<double>("PeakAmpRange", 2.);
-    fFloatBaseline = pset.get< bool >("FloatBaseline", false);
-
-    return;
-  }
+  : fMinWidth(pset.get<double>("MinWidth", 0.5)),
+    fMaxWidthMult(pset.get<double>("MaxWidthMult", 3.)),
+    fPeakRange(pset.get<double>("PeakRangeFact", 2.)),
+    fAmpRange(pset.get<double>("PeakAmpRange", 2.))
```
Comment (absolution1) [inline larreco/HitFinder/HitFinderTools/PeakFitterMrqdt_tool.cc:52]: default values

**Comment by ghost** [comment]
Yes, those warnings are unrelated to this pull request. The relevant DUNE CI tests are OK.

**Comment by yangtj207** [comment]
I agree with @chalt007.

**Review by absolution1 — COMMENTED** [review]
All looks fine to me.
I noticed that the initialisation of the FCL parameters is littered with default values, which are usually advised against.  It looks like those are actually passed through from the old code.

I'm wondering if those default values should be dropped; if an experiment misconfigures their FCL config, that experiment would pick up the default values and be unaware of their potentially dangerous error.

I see all of this code was originally implemented by @SFBayLaser so maybe he should comment on the default values.

**Comment by SFBayLaser** [comment]
Perhaps there is a better way but having default values helps people to get an idea of what is thought to be a reasonable value for a FHICL variiable in the code. Clearly not going to be correct for all experiments but I think the onus is on an experiment using a module/tools to have understood what the settings are and override them in their own definitions files. I inherited the module early in the MicroBooNE development, many of the default vallues were coming from Argneut and were not unreasonable for MicroBooNE (in the beginning). However, I can say many values are different for ICARUS. 
So if the collective wisdom is to remove the values then no worries from me but it would be worth having some mechanism in place to help give new users of the code some idea of what the values might be without having to look up some other experiment's FHICL files (which they may not have access to).

**Comment by absolution1** [comment]
I think code comments documenting what initial values to use would be roughly equivalent to communicating with default values in the initialisation.  Sure enough, code comments could become stale but so can the default values in the data member initialisation, I guess.

Thinking about this a bit more, it's probably beyond the scope of this pull request to change the status quo; we'd probably need a larger more general discussion about how to communicate reasonable starting 'guesses' for FCL parameters. I'm not sure whether default values are the way to go for the reason I outlined above.  Maybe let's save that conversation for another day (sorry for starting it!)

In regards to the code codes, the data member initialisations from FCL parameters are functionally equivalent to what they were before, which is good.  So, let's go ahead.

**Comment by absolution1** [comment]
approve

**Comment by lgarren** [comment]
approve

**Comment by lgarren** [comment]
@sabasehrish There are problems compiling this with c7.  Please fix.

**Comment by absolution1** [comment]
trigger build

**Comment by lgarren** [comment]
Thanks @sabasehrish   I confirm that this PR now compiles with c7.

**Comment by lgarren** [comment]
@absolution1  This PR looks good to me.  Do you have any concerns?

**Comment by absolution1** [comment]
The code itself looks fine but I noticed that some of the CI tests for both DUNE and SBND's SL6:py2 builds are reporting warning on some tests, specifically a change in product size:

DUNE: https://dbweb5.fnal.gov:8443/LarCI/app/ns:dune/build_detail/phase_details?build_id=dune_ci/9042&platform=Linux%202.6.32-754.28.1.el6.x86_64&phase=ci_tests&buildtype=slf6%20e19:py2:prof

SBND: https://dbweb5.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/4769&platform=Linux%202.6.32-754.28.1.el6.x86_64&phase=ci_tests&buildtype=slf6%20e19:py2:prof

Weirdly the SL7 builds for both experiments report no warnings.

@yangtj207 @chalt007 is this expected in the DUNE case?

**Comment by absolution1** [comment]
Actually, looking again: it looks like the DUNE and SBND SL6 builds have reported similar warnings for quite a few separate CI tests now so its most likely unrelated to this pull request.  Let's proceed.

**Comment by absolution1** [comment]
approve

**Comment by lgarren** [comment]
approve

**Pull request merged**